src/mrpro/operators/_CartesianSamplingOp.py

```diff
@@ -0,0 +1,172 @@
+            raise ValueError('k-space data shape missmatch')
+
+        if len(self._fft_dims) > 0 and len(self._fft_idx) > 0:
+            if len(self._fft_idx_full) == 0 or self._fft_idx_full.shape[1] != y.shape[1]:
+                coil_idx = torch.ones(
+                    *(y.shape[:2] + self._fft_idx.shape[2:]), dtype=torch.int64, device=y.device
+                ) * rearrange(
+                    torch.linspace(0, y.shape[1] - 1, y.shape[1], dtype=torch.int64, device=y.device),
+                    'coils->1 coils 1 1 1',
+                )
+                other_idx = torch.ones_like(self._fft_idx, dtype=torch.int64, device=y.device) * rearrange(
+                    torch.linspace(0, y.shape[0] - 1, y.shape[0], dtype=torch.int64, device=y.device),
+                    'other->other 1 1 1 1',
+                )
+                self._fft_idx_full = (
+                    self._fft_idx
+                    + other_idx * y.shape[1]
+                    + coil_idx * self._encoding_shape.z * self._encoding_shape.y * self._encoding_shape.x
+                )
+
+            return torch.take(y, self._fft_idx_full)
+        else:
+            return y
+
+    def adjoint(self, y: torch.Tensor) -> torch.Tensor:
+        """Adjoint operator sorting data into the encoding_space matrix.
+
+        Parameters
+        ----------
+        y
+            k-space data in original (i.e. acquired shape)
+
+        Returns
+        -------
+            k-space data sorted into encoding_space matrix
+        """
+
+        if self._kshape[1:] != y.shape[-3:]:
+            raise ValueError('k-space data shape missmatch')
+
+        if len(self._fft_dims) > 0 and len(self._fft_idx) > 0:
+            if len(self._fft_idx_full) == 0 or self._fft_idx_full.shape[1] != y.shape[1]:
+                coil_idx = torch.ones(
+                    *(y.shape[:2] + self._fft_idx.shape[2:]), dtype=torch.int64, device=y.device
+                ) * rearrange(
+                    torch.linspace(0, y.shape[1] - 1, y.shape[1], dtype=torch.int64, device=y.device),
+                    'coils->1 coils 1 1 1',
+                )
+                other_idx = torch.ones_like(self._fft_idx, dtype=torch.int64, device=y.device) * rearrange(
+                    torch.linspace(0, y.shape[0] - 1, y.shape[0], dtype=torch.int64, device=y.device),
+                    'other->other 1 1 1 1',
+                )
+                self._fft_idx_full = (
+                    self._fft_idx
+                    + other_idx * y.shape[1]
+                    + coil_idx * self._encoding_shape.z * self._encoding_shape.y * self._encoding_shape.x
+                )
+
+            ysort = torch.zeros(
+                *(y.shape[:2] + (self._encoding_shape.z, self._encoding_shape.y, self._encoding_shape.x)),
+                dtype=y.dtype,
+                device=y.device,
+            )
+
+            return ysort.put_(self._fft_idx_full, y, accumulate=True)
```

Comment:
I would chainthis directly to the zeros call above.
As only after the put `ysorted `is y sorted, and to signify this is one of the exceptions where in place operations are ok: on newly created tensors that will never require gradients.